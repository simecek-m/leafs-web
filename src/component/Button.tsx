import { FC, ButtonHTMLAttributes } from "react";
import clsx from "clsx";
import { IconProp } from "@fortawesome/fontawesome-svg-core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

type ButtonVariant = "primary" | "destructive";

const ButtonVariants: Record<ButtonVariant, string> = {
  primary: "bg-primary-700",
  destructive: "bg-red-600",
};

export interface ButtonProps {
  type?: ButtonHTMLAttributes<HTMLButtonElement>["type"];
  disabled?: boolean;
  isSubmitting?: boolean;
  onClick?: () => void;
  children: string;
  icon: IconProp;
  hoverIcon: IconProp;
  variant?: ButtonVariant;
}

const Button: FC<ButtonProps> = ({
  onClick,
  type = "button",
  disabled = false,
  isSubmitting = false,
  icon,
  variant = "primary",
  hoverIcon,
  children,
}) => {
  return (
    <button
      onClick={onClick}
      type={type}
      disabled={disabled}
      className={clsx(
        { "cursor-wait": isSubmitting },
        ButtonVariants[variant],
        "group flex w-full flex-row items-center justify-center gap-2 overflow-hidden rounded-full py-2 px-2 pr-5 font-bold text-white disabled:cursor-not-allowed disabled:opacity-40 sm:w-fit"
      )}
    >
      {!disabled && (
        <div className="relative flex">
          {icon && (
            <FontAwesomeIcon
              icon={icon}
              className="h-4 w-4 translate-y-0 rounded-full bg-white bg-opacity-25 p-2 text-white transition duration-500 ease-in-out group-hover:-translate-y-10"
            />
          )}
          {hoverIcon && (
            <FontAwesomeIcon
              icon={hoverIcon}
              className="absolute h-4 w-4 translate-y-10 rounded-full bg-white bg-opacity-25 p-2 text-white transition duration-500 ease-in-out group-hover:translate-y-0"
            />
          )}
        </div>
      )}
      {disabled && (
        <FontAwesomeIcon
          icon="xmark"
          className="h-4 w-4 rounded-full bg-white bg-opacity-25 p-2 text-white"
        />
      )}
      <span className="">{isSubmitting ? "processing" : children}</span>
    </button>
  );
};

export default Button;
