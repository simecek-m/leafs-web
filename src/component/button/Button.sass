.button
  padding: 15px 30px
  color: var(--on-surface)
  background: var(--transparent)
  outline: none
  border: none
  font-size: 1.2rem
  text-transform: lowercase
  font-weight: bold
  cursor: pointer
  transition: 0.5s ease all
  box-shadow: 0.3em 0.3em 0 0 var(--brand), inset 0.3em 0.3em 0 0 var(--brand)

  &:hover
    color: var(--on-brand)
    box-shadow: 0 0 0 0 var(--brand), inset 6em 3.5em 0 0 var(--brand)
