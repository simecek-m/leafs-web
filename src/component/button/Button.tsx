import { IconProp } from "@fortawesome/fontawesome-svg-core";
import clsx from "clsx";
import ButtonIcon from "component/button/ButtonIcon";
import { ButtonHTMLAttributes, FC } from "react";
import { ActionType } from "type";

const ButtonVariants: Record<ActionType, string> = {
  primary:
    "border-primary-700 dark:border-primary-300 text-primary-700 dark:text-primary-300",
  error: "border-red-700 dark:border-red-300 text-red-700 dark:text-red-300",
};

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  icon: IconProp;
  hoverIcon?: IconProp;
  variant?: ActionType;
  isDisabled?: boolean;
  isSubmitting?: boolean;
  onClick?: () => void;
}

const Button: FC<ButtonProps> = ({
  onClick,
  type = "button",
  isDisabled = false,
  isSubmitting = false,
  icon,
  variant = "primary",
  hoverIcon,
  children,
}) => {
  return (
    <button
      onClick={onClick}
      type={type}
      disabled={isDisabled || isSubmitting}
      className={clsx(
        "group flex w-full shrink-0 flex-row items-center justify-center gap-2 overflow-hidden rounded-full border-4 bg-transparent py-2 px-2 pr-5 font-bold sm:w-fit",
        { "cursor-wait": isSubmitting },
        { "cursor-not-allowed opacity-60": isDisabled },
        ButtonVariants[variant]
      )}
    >
      <ButtonIcon
        icon={icon}
        variant={variant}
        hoverIcon={hoverIcon}
        isDisabled={isDisabled}
        isSubmitting={isSubmitting}
      />
      <span
        className={clsx("capitalize", {
          "font-normal italic": !!isSubmitting && !isDisabled,
        })}
      >
        {isSubmitting && !isDisabled ? "processing" : children}
      </span>
    </button>
  );
};

export default Button;
