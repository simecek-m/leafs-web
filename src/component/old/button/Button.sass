@import "colors.sass"

.button-component
  display: inline-block
  color: $text-primary
  border: 2px solid $text-primary
  background: none
  padding: 10px 30px
  border-radius: 5px
  user-select: none
  cursor: pointer
  transition: 0.3s ease all
  font-size: 20px
  font-weight: bold
  margin: 20px 0

  .button-component-icon
    display: inline-block
    margin-right: 20px

  &:hover
    color: $background-primary
    border: 2px solid $text-primary
    background: $text-primary

  &.primary
    color: $brand-color
    border-color: $brand-color

    &:hover
      color: $background-primary
      border: 2px solid $brand-color
      background: $brand-color

  &.info
    color: $info-color
    border-color: $info-color

    &:hover
      color: $background-primary
      border: 2px solid $info-color
      background: $info-color

  &.success
    color: $success-color
    border-color: $success-color

    &:hover
      color: $background-primary
      border: 2px solid $success-color
      background: $success-color

  &.warn
    color: $warn-color
    border-color: $warn-color

    &:hover
      color: $background-primary
      border: 2px solid $warn-color
      background: $warn-color

  &.error
    color: $error-color
    border-color: $error-color

    &:hover
      color: $background-primary
      border: 2px solid $error-color
      background: $error-color
