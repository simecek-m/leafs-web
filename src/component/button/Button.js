import "component/button/Button.sass";

export default function Button({ onClick, children }) {
  return (
    <button onClick={onClick} className="button">
      {children}
    </button>
  );
}
