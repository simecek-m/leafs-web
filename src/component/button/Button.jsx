import { func } from "prop-types";
import "component/button/Button.sass";

export default function Button({ onClick, children }) {
  return (
    <button onClick={onClick} className="button">
      {children}
    </button>
  );
}

Button.propTypes = {
  onClick: func.isRequired,
};
