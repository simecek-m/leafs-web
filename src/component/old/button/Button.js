import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import "component/old/button/Button.sass";

function Button({
  icon = null,
  text = "Button",
  type = "default",
  onClick = null
}) {
  const buttonIcon = icon ? (
    <FontAwesomeIcon className="button-component-icon" icon={icon} />
  ) : null;
  return (
    <span className={`button-component ${type}`} onClick={onClick}>
      {buttonIcon}
      <span className="text">{text.toUpperCase()}</span>
    </span>
  );
}

export default Button;
