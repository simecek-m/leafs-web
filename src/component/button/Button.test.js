import React from "react";
import Button from "component/button/Button";
import { faDatabase } from "@fortawesome/free-solid-svg-icons";
import { shallow } from "enzyme";

const TEST_BUTTON_CUSTOM_TYPE = "warn";
const TEST_BUTTON_CUSTOM_TEXT = "send";
const TEST_BUTTON_CUSTOM_ICON = faDatabase;
const TEST_BUTTON_CUSTOM_CALLBACK = () => {};

describe("Button component", () => {
  test("should render default Button", () => {
    const wrapper = shallow(<Button />);
    expect(wrapper).toMatchSnapshot();
  });

  test("should render custom Button", () => {
    const wrapper = shallow(
      <Button
        icon={TEST_BUTTON_CUSTOM_ICON}
        text={TEST_BUTTON_CUSTOM_TEXT}
        onClick={TEST_BUTTON_CUSTOM_CALLBACK}
        type={TEST_BUTTON_CUSTOM_TYPE}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
