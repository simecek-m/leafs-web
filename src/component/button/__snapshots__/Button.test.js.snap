// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render custom Button 1`] = `
<span
  className="button-component warn"
  onClick={[Function]}
>
  <FontAwesomeIcon
    className="button-component-icon"
    icon={
      Object {
        "icon": Array [
          448,
          512,
          Array [],
          "f1c0",
          "M448 73.143v45.714C448 159.143 347.667 192 224 192S0 159.143 0 118.857V73.143C0 32.857 100.333 0 224 0s224 32.857 224 73.143zM448 176v102.857C448 319.143 347.667 352 224 352S0 319.143 0 278.857V176c48.125 33.143 136.208 48.572 224 48.572S399.874 209.143 448 176zm0 160v102.857C448 479.143 347.667 512 224 512S0 479.143 0 438.857V336c48.125 33.143 136.208 48.572 224 48.572S399.874 369.143 448 336z",
        ],
        "iconName": "database",
        "prefix": "fas",
      }
    }
  />
  <span
    className="text"
  >
    SEND
  </span>
</span>
`;

exports[`Button component should render default Button 1`] = `
<span
  className="button-component default"
  onClick={null}
>
  <span
    className="text"
  >
    BUTTON
  </span>
</span>
`;
